From e9a95db1f8c50895e4f0de33f3a5aecaaffc7a49 Mon Sep 17 00:00:00 2001
From: Ed Tanous <etanous@nvidia.com>
Date: Sat, 3 Aug 2024 04:15:17 -0700
Subject: [PATCH] Add phy hacks

GB200NVL BMC needs to set 600 mv drive strength on the Realtek part.
Upstream-Status: Pending [Not submitted to upstream yet]
---
 drivers/net/phy/Kconfig   |  6 ++++++
 drivers/net/phy/realtek.c | 22 +++++++++++++++++++++-
 2 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/drivers/net/phy/Kconfig b/drivers/net/phy/Kconfig
index 2c739f73b30c..4c7b90c947d7 100644
--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -487,3 +487,9 @@ endif # PHYLIB
 config MICREL_KS8995MA
 	tristate "Micrel KS8995MA 5-ports 10/100 managed Ethernet switch"
 	depends on SPI
+
+config REALTEK_600MV_DRIVE_STRENGTH
+	bool "600Mv Drive strength"
+	help
+		Enables 600Mv Drive strength
+
diff --git a/drivers/net/phy/realtek.c b/drivers/net/phy/realtek.c
index e4a2caec0388..e93adee6cc64 100644
--- a/drivers/net/phy/realtek.c
+++ b/drivers/net/phy/realtek.c
@@ -119,6 +119,24 @@ static int rtl821x_write_page(struct phy_device *phydev, int page)
 	return __phy_write(phydev, RTL821x_PAGE_SELECT, page);
 }
 
+static void set_600mv_drive_strength(struct phy_device *phydev){
+	int ret;
+	struct device *dev = &phydev->mdio.dev;
+	dev_info(dev, "Writing drive strength indirect \n");
+	rtl821x_write_page(phydev, 0xa43);
+	ret = __phy_write(phydev, 0x1b, 0xdcd0);
+	rtl821x_write_page(phydev, 0xa43);
+	ret = __phy_write(phydev, 0x1c, 0x1096);
+	rtl821x_write_page(phydev, 0xa43);
+	ret = __phy_write(phydev, 0x1b, 0xdcd2);
+	rtl821x_write_page(phydev, 0xa43);
+	ret = __phy_write(phydev, 0x1c, 0xB490);
+	rtl821x_write_page(phydev, 0xa43);
+	ret = __phy_write(phydev, 0x1f, 0x0);
+
+	dev_info(dev, "Done writing drive strength indirect\n");
+}
+
 static int rtl821x_probe(struct phy_device *phydev)
 {
 	struct device *dev = &phydev->mdio.dev;
@@ -155,7 +173,9 @@ static int rtl821x_probe(struct phy_device *phydev)
 	}
 
 	phydev->priv = priv;
-
+#ifdef CONFIG_REALTEK_600MV_DRIVE_STRENGTH
+	set_600mv_drive_strength(phydev);
+#endif
 	return 0;
 }
 
-- 
2.43.0

